[jenkins] improve revsion, date and branch name detection

Rechi opened the pull request:
<!--- Provide a general summary of your change in the Title above -->

## Description
- more exact check for pull request in attached state
- append head or merge for a pull request in attached state
- favour branch `--points-at` over branch `--contains` in detached state
- detect pull requests in detached state
- replace slashes in branch names with a dash
- enlarge commit hash as it sometimes isn't distinct anymore
- ~output date in wanted format from git~ (git version on osx build nodes is to old)
<!--- Describe your change in detail -->

## Motivation and Context
Fix file names on the mirror for PR test builds.
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
on osx and windows for branches (+pr branches) in non detached and detached state
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

<!--## Screenshots (if appropriate):-->

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Improvement (non-breaking change which improves existing functionality)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

Memphiz reviewed (approved):
Not to easy to grasp what the changes do but i think we will see it soon if there are issues with uploaded filenames. Ok from my side.

Rechi commented:
The major change is the correct naming of jenkins PR builds. Without this change they don't have a branch in the filename. When using `Project BuildMulti-PR-Manually` one doesn't know what build it actually was.
The other changes are just minor improvements.

I checked the jenkins log of this build for correct file names on all platforms.

Rechi committed:
[buildsteps][windows] run-tests doesn't use branch name

Rechi committed:
[jenkins] improve revsion, date and branch name detection

- more exact check for pull request in attached state
- append head or merge for a pull request in attached state
- favour branch --points-at over branch --contains in detached state
- detect pull requests in detached state
- replace slashes in branch names with a dash
- enlarge commit hash as it sometimes isn't distinct anymore

Rechi commented:
Floating point exception in Linux test suite is unrelated.
